compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -0,0 +1,242 @@
+        DefKind::Impl { of_trait: false } => FnKind::AssocInherentImpl,
+        _ => FnKind::Free,
+    }
+}
+
+// Lifetime parameters must be declared before type and const parameters.
+// Therefore, When delegating from a free function to a associated function,
+// generic parameters need to be reordered:
+//
+// trait Trait<'a, A> {
+//     fn foo<'b, B>(...) {...}
+// }
+//
+// reuse Trait::foo;
+// desugaring:
+// fn foo<'a, 'b, This: Trait<'a, A>, A, B>(...) {
+//     Trait::foo(...)
+// }
+fn create_remap_table<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId, sig_id: DefId) -> RemapTable {
+    let caller_generics = tcx.generics_of(def_id);
+    let callee_generics = tcx.generics_of(sig_id);
+    let mut remap_table: RemapTable = FxHashMap::default();
+    for caller_param in &caller_generics.own_params {
+        let callee_index = callee_generics.param_def_id_to_index(tcx, caller_param.def_id).unwrap();
+        remap_table.insert(callee_index, caller_param.index);
+    }
+    remap_table
+}
+
+pub(crate) fn inherit_generics_for_delegation_item<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: LocalDefId,
+    sig_id: DefId,
+) -> Option<ty::Generics> {
+    let caller_kind = fn_kind(tcx, def_id.into());
+    let callee_kind = fn_kind(tcx, sig_id);
+
+    // FIXME(fn_delegation): Support generics on associated delegation items.
+    // Error was reported in `check_delegation_constraints`.
+    match (caller_kind, callee_kind) {
+        (FnKind::Free, _) => {
+            let mut own_params = vec![];
+
+            let callee_generics = tcx.generics_of(sig_id);
+            if let Some(parent_sig_id) = callee_generics.parent {
+                let parent_sig_generics = tcx.generics_of(parent_sig_id);
+                own_params.append(&mut parent_sig_generics.own_params.clone());
+            }
+            own_params.append(&mut callee_generics.own_params.clone());
+
+            // lifetimes go first
+            own_params.sort_by_key(|key| key.kind.is_ty_or_const());
+
+            for (idx, param) in own_params.iter_mut().enumerate() {
+                param.index = idx as u32;
+                // Default type parameters are not inherited: they are not allowed
+                // in fn's.
+                if let ty::GenericParamDefKind::Type { synthetic, .. } = param.kind {
+                    param.kind = ty::GenericParamDefKind::Type { has_default: false, synthetic }
+                }
+            }
+
+            let param_def_id_to_index =
+                own_params.iter().map(|param| (param.def_id, param.index)).collect();
+
+            Some(ty::Generics {
+                parent: None,
+                parent_count: 0,
+                own_params,
+                param_def_id_to_index,
+                has_self: false,
+                has_late_bound_regions: callee_generics.has_late_bound_regions,
+                host_effect_index: None,
+            })
+        }
+        _ => None,
```

Comment:
Can use `span_delayed_bug` here if an error is reported earlier.
Same may apply to other matches below.

Reply:
`None` does not always mean an error. Callee may not have parameters, in this case "normal flow" will be taken. See `predicated_of`, `generics_of`.